planner/cascades/optimize.go

```diff
@@ -31,21 +31,23 @@ var DefaultOptimizer = NewOptimizer()
 type Optimizer struct {
 	transformationRuleMap map[memo.Operand][]Transformation
 	implementationRuleMap map[memo.Operand][]ImplementationRule
+	patternMap            map[Transformation]*memo.Pattern
```

Comment:
What about adding a ID field to pattern and changing the map to `[]*memoPattern`?